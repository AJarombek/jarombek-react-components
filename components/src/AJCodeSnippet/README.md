### Overview

Reusable component that displayed programming language code snippets with syntax highlighting 
provided by `highlight.js`.

**Use cases:**

```javascript
import { AJCodeSnippet } from 'jarombek-react-components';

const CustomComponent = () => 
  <AJCodeSnippet language="javascript">
    const j = 25;
  </AJCodeSnippet>
```

### Files

| Filename                   | Description                                                                           |
|----------------------------|---------------------------------------------------------------------------------------|
| `AJCodeSnippet.js`         | The reusable `AJCodeSnippet` component.                                               |
| `index.js`                 | Index file which exports the `AJCodeSnippet` component.                               |
| `styles.js`                | JSS styles for the `AJCodeSnippet` component.                                         |
