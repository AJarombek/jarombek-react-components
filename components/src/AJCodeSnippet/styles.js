/**
 * Stylesheet for the AJCodeSnippet component using JSS.
 * Author: Andrew Jarombek
 * Date: 12/30/2019
 */

export default {
  hljs: {
    display: 'block',
    overflowX: 'auto',
    padding: '16px 8px 16px 16px',
    color: '#ddd',
    fontSize: '18px',
    whiteSpace: 'pre',
    lineHeight: 1.4,
    borderRadius: '2px',
    backgroundColor: '#333',
    transition: 'all .4s ease',
    boxShadow: '0 1px 3px rgba(170, 170, 170, 0.1)',
    '-webkit-overflow-scrolling': 'touch',
    '-webkit-transition': 'revert',

    '&:hover': {
      boxShadow: '0 1px 3px rgba(170, 170, 170, 0.8)',
    },

    '&:before': {
      content: 'attr(title)',
      display: 'block',
      fontSize: '14px',
      color: '#aaa',
      position: 'sticky',
      left: '90%',
      float: 'right',

      [`@media screen and (max-width: 700px)`]: {
        marginRight: '5px',
        left: '88%',
      },

      [`@media screen and (max-width: 480px)`]: {
        marginRight: '10px',
        left: '85%',
      },
    },

    '& .hljs-comment, & .hljs-quote': {
      color: '#5c6370',
      fontStyle: 'italic',
    },

    '& .hljs-doctag, & .hljs-keyword, & .hljs-formula': {
      color: '#c678dd',
    },

    '& .hljs-section, & .hljs-name, & .hljs-selector-tag, & .hljs-deletion, & .hljs-subst': {
      color: '#e06c75',
    },

    '& .hljs-literal': {
      color: '#56b6c2',
    },

    '& .hljs-string, & .hljs-regexp, & .hljs-addition, & .hljs-attribute, & .hljs-meta-string': {
      color: '#98c379',
    },

    '& .hljs-built_in, & .hljs-class .hljs-title': {
      color: '#e6c07b',
    },

    '& .hljs-attr, & .hljs-variable, & .hljs-template-variable, & .hljs-type': {
      color: '#d19a66',
    },
    '& .hljs-selector-class, & .hljs-selector-attr, & .hljs-selector-pseudo, & .hljs-number': {
      color: '#d19a66',
    },

    '& .hljs-symbol, & .hljs-bullet, & .hljs-link, & .hljs-meta, & .hljs-selector-id': {
      color: '#61aeee',
    },
    '& .hljs-title': {
      color: '#61aeee',
    },

    '& .hljs-emphasis': {
      fontStyle: 'italic',
    },

    '& .hljs-strong': {
      fontWeight: 'bold',
    },

    '& .hljs-link': {
      textDecoration: 'underline',
    },
  },
};
