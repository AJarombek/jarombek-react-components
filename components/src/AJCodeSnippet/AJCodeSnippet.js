/**
 * Reusable code snippet component for displaying code with syntax highlighting.
 * @author Andrew Jarombek
 * @since 12/30/2019
 */

import React, {useEffect} from 'react';
import PropTypes from 'prop-types';
import {createUseStyles} from 'react-jss';
import hljs from 'highlight.js';
import classnames from 'classnames';

import styles from './styles';

const useStyles = createUseStyles(styles);

/**
 * Component representing a snippet of code from one of multiple different languages.  The code is
 * syntax highlighted with the help of highlight.js.
 * @param children The code the is displayed in the code snippet.
 * @param language The language that the code is written in.  The language determines how the code
 * is syntax highlighted.
 * @param className Custom class attribute(s) attached to the component.
 * @return {*} React elements that represent a snippet of code.
 * @constructor
 */
const AJCodeSnippet = ({children, language, className}) => {
  const classes = useStyles();

  const languages = {
    javascript: {
      title: "JavaScript",
      name: "javascript",
      className: "javascript"
    },
    java: {
      title: "Java",
      name: "java",
      className: "java"
    }
  };

  const lang = languages[language] || languages.javascript;

  if (!language) {
    lang.title = "";
  }

  let highlightedCode;

  useEffect(() => {
    hljs.initHighlightingOnLoad();
    hljs.highlightBlock(highlightedCode);
  });

  return (
    <div className={classnames("aj-code-snippet", className)}>
      <code title={lang.title} className={classnames(lang.className, classes.hljs)}
            ref={element => highlightedCode = element}>
        {children}
      </code>
    </div>
  );
};

AJCodeSnippet.propTypes = {
  children: PropTypes.node.isRequired,
  language: PropTypes.oneOf(['java', 'javascript']),
  className: PropTypes.string,
};

export default AJCodeSnippet;
