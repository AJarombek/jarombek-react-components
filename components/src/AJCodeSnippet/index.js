import AJCodeSnippet from './AJCodeSnippet';

export default AJCodeSnippet;
