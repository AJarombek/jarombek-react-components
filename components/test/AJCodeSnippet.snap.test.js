import React from 'react';
import AJCodeSnippet from '../src/AJCodeSnippet/AJCodeSnippet';
import renderer from 'react-test-renderer';

/**
 * Snapshot test for the {@link ../../components/AJCodeSnippet} component.
 * @author Andrew Jarombek
 * @since 12/19/2019
 */

it('renders correctly', () => {
  const tree = renderer.create(
    <AJCodeSnippet language="javascript">
      {`
      const list = [0, 1, 2, 3];
      for (let item in list) {
        console.info(item);
      }
      `}
    </AJCodeSnippet>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
