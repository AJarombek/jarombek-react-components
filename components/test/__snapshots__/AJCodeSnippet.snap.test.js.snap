// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <code
    className="javascript hljs-0-2-1"
    title="JavaScript"
  >
    
      const list = [0, 1, 2, 3];
      for (let item in list) {
        console.info(item);
      }
      
  </code>
</div>
`;
